2020/day-5-binary-boarding/index.ts

```diff
@@ -0,0 +1,71 @@
+const firstHalf = (arr: any[]) => arr.slice(0, arr.length / 2)
+const lastHalf = (arr: any[]) => arr.slice(arr.length / 2)
+
+type State = {
+  row: number[]
+  col: number[]
+}
+
+type Char = 'F' | 'B' | 'L' | 'R'
+
+const reducer = (state: State, char: Char) => {
+  switch (char) {
+    case 'F': {
+      return {
+        ...state,
+        row: firstHalf(state.row),
+      }
+    }
+
+    case 'B': {
+      return {
+        ...state,
+        row: lastHalf(state.row),
+      }
+    }
+
+    case 'L': {
+      return {
+        ...state,
+        col: firstHalf(state.col),
+      }
+    }
+
+    case 'R': {
+      return {
+        ...state,
+        col: lastHalf(state.col),
+      }
+    }
+
+    default: {
+      throw new Error('Unknown code character')
+    }
+  }
+}
+
+export const findSeatPosition = (code: string) => {
+  const totalRows = Array.from(Array(128), (_, i) => i)
+  const totalCols = Array.from(Array(8), (_, i) => i)
+
+  let state = {
+    row: totalRows,
+    col: totalCols,
+  }
+
+  code.split('').forEach(char => {
+    state = reducer(state, char as Char)
+  })
```

Comment:
This feels like a reduce function to me?

```suggestion
  code.split('').reduce((acc, char) => reducer(acc, char as Char), state)
```

Reply:
ooh, that's pretty shiny. Hadn't thought of that. Let's do it